Web Inspector: Support editing @rules in the Styles sidebar

patrickangle opened the pull request:
#### 33803c89a12929f179cf29e76afc037e6c5fdf85
<pre>
Web Inspector: Support editing @rules in the Styles sidebar
<a href="https://bugs.webkit.org/show_bug.cgi?id=246768">https://bugs.webkit.org/show_bug.cgi?id=246768</a>
rdar://99871652

Reviewed by Devin Rousso.

This work is best thought of in three interlocking pieces:
- Rework chunks of InspectorStyleSheet to allow us to determine source data for @rules, including the header range,
which is the range of the text between `@media` (and friends) and the opening curly bracket for the rule set.
- Generalizing setting a rule&apos;s selector to just be setting its header text so we can do so with non-style rules.
- Implement frontend support for editing these rules in the same way selectors are currently editable.

* LayoutTests/inspector/css/setGroupingHeaderText-expected.txt: Added.
* LayoutTests/inspector/css/setGroupingHeaderText.html: Added.

* LayoutTests/inspector/css/generateFormattedText-expected.txt:
* LayoutTests/inspector/css/getMatchedStylesForNode-expected.txt:
- Rebaseline expectations for new CSS.Grouping properties.

* Source/JavaScriptCore/inspector/protocol/CSS.json:
- Add rule ID to CSS.Grouping so that we can use it later to edit the header text of the grouping.
- Add method CSS.setGroupingHeaderText to change a CSS.Grouping&apos;s header text.

* Source/WebCore/css/parser/CSSParser.h:
* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::CSSParserImpl::consumeContainerRule):
- Provide a correct header range by keeping a copy of the original prelude before consumption to use for informing the
observer.

* Source/WebCore/css/parser/CSSParserImpl.h:
- Expose `consumeAtRule` publicly for use in InspectorStyleSheet

* Source/WebCore/inspector/InspectorStyleSheet.cpp:
(WebCore::flatteningStrategyForStyleRuleType):
- Move the logic for determining how we unflatten rules to a shared static method so that the source data and flat CSSOM
rule lists always consider the same set of rules in the same way.

(WebCore::parserContextForDocument):
(WebCore::isValidRuleHeaderText):
(WebCore::protocolGroupingTypeForStyleRuleType):
- Abstract logic that relies on rule type into static methods so that the various bits that need implemented when adding
an editable rule type are in close proximity to each other.

(WebCore::flattenSourceData):
(WebCore::ParsedStyleSheet::setSourceData):
- Use `flatteningStrategyForStyleRuleType` to determine how we should handle each piece of source data.

(WebCore::StyleSheetHandler::startRuleHeader):
(WebCore::StyleSheetHandler::setRuleHeaderEnd):
- Fix to never produce a range end that occurs before the range&apos;s start.

(WebCore::sourceURLForCSSRule):
(WebCore::InspectorStyleSheet::buildObjectForGrouping):
(WebCore::InspectorStyleSheet::buildArrayForGroupings):
- Split building the grouping array into a few discrete, reusable pieces. For the four types of rules we are making
editable, we will use `buildObjectForGrouping` to build those rule objects. For other real that are currently turned
into fake groupings, we leave the logic mostly intact (which means they are not editable).

(WebCore::InspectorStyleSheet::ruleHeaderText):
(WebCore::InspectorStyleSheet::setRuleHeaderText):
(WebCore::InspectorStyleSheet::ruleSelector): Deleted.
(WebCore::isValidSelectorListString): Deleted.
(WebCore::InspectorStyleSheet::setRuleSelector): Deleted.
- Setting a rule&apos;s header text and setting a rule&apos;s selector(s) are (except for some optimizations) the same operation.
For CSSStyleRule we still follow the existing logic of modifying a selector via CSSOM, which saves us from reparsing the
entire style sheet. For other rules, this is more complicated because our CSSOM implementation doesn&apos;t allow mutating
the header text, and changing these rules could change which rules apply to which nodes beyond just the rule we are
mutating.

(WebCore::InspectorStyleSheet::deleteRule):
(WebCore::InspectorStyleSheet::ruleForId const):
(WebCore::InspectorStyleSheet::cssStyleRulesSplitFromSameRule):
(WebCore::InspectorStyleSheet::buildObjectForSelectorList):
(WebCore::InspectorStyleSheet::buildObjectForRule):
(WebCore::InspectorStyleSheet::styleForId const):
(WebCore::InspectorStyleSheet::ruleOrStyleId const):
(WebCore::InspectorStyleSheet::ruleSourceDataFor const):
(WebCore::InspectorStyleSheet::ruleIndexByStyle const):
(WebCore::InspectorStyleSheet::buildArrayForRuleList):
(WebCore::InspectorStyleSheet::collectFlatRules):
(WebCore::InspectorStyleSheet::ruleId const): Deleted.
- Don&apos;t assume that a CSSRule will always be a CSSStyleRule.
- Allow both styles and rules to be used to look up the ID for a rule/style.

* Source/WebCore/inspector/InspectorStyleSheet.h:
(WebCore::InspectorStyleSheet::canBind const):
(WebCore::InspectorStyleSheet::styleId const): Deleted.

* Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
(WebCore::InspectorCSSAgent::setRuleSelector):
(WebCore::InspectorCSSAgent::setGroupingHeaderText):
* Source/WebCore/inspector/agents/InspectorCSSAgent.h:
- Generalize setting selectors to also support setting header text of @rules.

(WebCore::InspectorCSSAgent::addRule):
- Don&apos;t assume that all rules are CSSStyleRules.

(WebCore::InspectorCSSAgent::asCSSStyleRule): Deleted.
- Remove helper that was only used in a few places and is better served by a dynamicDowncast anyways.

* Source/WebInspectorUI/UserInterface/Main.html:

* Source/WebInspectorUI/UserInterface/Models/CSSGrouping.js:
(WI.CSSGrouping.prototype.get ownerStyleSheet):
(WI.CSSGrouping.prototype.get id):
(WI.CSSGrouping.prototype.get type):
(WI.CSSGrouping.prototype.get editable):
(WI.CSSGrouping.prototype.get text):
- Add necessary information for editing, which is mostly keeping track of our nodeStyles and ownerStyleSheet like style
rules do, and accepting an optional ID, which when present allows us to perform edits.

(WI.CSSGrouping.prototype.async setText):
- Handle setting the text for the grouping. For groupings that no longer apply after modification, we need to update the
current grouping object so that it is accurate to the current state, as it will remain visible to the user until they
switch to a different node.

* Source/WebInspectorUI/UserInterface/Models/DOMNodeStyles.js:
(WI.DOMNodeStyles):
(WI.DOMNodeStyles.prototype.set ignoreNextContentDidChangeForStyleSheet):
(WI.DOMNodeStyles.prototype.refresh.fetchedMatchedStyles):
(WI.DOMNodeStyles.prototype._parseRulePayload):
- Parse the additional data in the payload to support editing.
- Use a shared representation of CSSGrouping&apos;s with an ID so that we can update the common representation during an
edit, thus allowing that same rule to update in all the locations it is visible for the current node. (It is not
necessary to update for other nodes, since they will have been invalidated by the stylesheet change and will reload
their styles.)

* Source/WebInspectorUI/UserInterface/Views/SpreadsheetCSSStyleDeclarationSection.css:
(.spreadsheet-css-declaration :is(.selector, .grouping).spreadsheet-rule-header-field):
(.spreadsheet-css-declaration :is(.selector, .grouping).spreadsheet-rule-header-field.editing):
- Style the grouping editor field the same way we style the selector editor field.

* Source/WebInspectorUI/UserInterface/Views/SpreadsheetCSSStyleDeclarationSection.js:
(WI.SpreadsheetCSSStyleDeclarationSection):
(WI.SpreadsheetCSSStyleDeclarationSection.prototype.initialLayout):
- Remove logic that lays out the CSS groupings, since those are dynamic and may change on future updates.

(WI.SpreadsheetCSSStyleDeclarationSection.prototype.layout):
(WI.SpreadsheetCSSStyleDeclarationSection.prototype._layoutGroupings.this._groupingElements.groupings.map.):
(WI.SpreadsheetCSSStyleDeclarationSection.prototype._layoutGroupings):
- Update groupings when layout is called to get the latest information from DOMNodeStyles.
- No longer merge together adjacent @rules like multiple @media or @layer rules, as those rules are editable now, and we
need to accurately represent the author&apos;s rules.
- Associate a text field with the grouping header text when applicable to enable editing.

(WI.SpreadsheetCSSStyleDeclarationSection.prototype.spreadsheetRuleHeaderFieldDidCommit):
(WI.SpreadsheetCSSStyleDeclarationSection.prototype.spreadsheetRuleHeaderFieldWillNavigate):
(WI.SpreadsheetCSSStyleDeclarationSection.prototype.spreadsheetRuleHeaderFieldDidDiscard):
(WI.SpreadsheetCSSStyleDeclarationSection.prototype.spreadsheetSelectorFieldDidCommit): Deleted.
(WI.SpreadsheetCSSStyleDeclarationSection.prototype.spreadsheetSelectorFieldWillNavigate): Deleted.
(WI.SpreadsheetCSSStyleDeclarationSection.prototype.spreadsheetSelectorFieldDidDiscard): Deleted.
- Share the delegate methods for the selector and the grouping fields.

(WI.SpreadsheetCSSStyleDeclarationSection.prototype._handleSpreadsheetSelectorFieldStartedEditing):
(WI.SpreadsheetCSSStyleDeclarationSection.prototype._handleSpreadsheetSelectorFieldStoppedEditing):
(WI.SpreadsheetCSSStyleDeclarationSection.prototype._handleSpreadsheetSelectorFieldDidCommit):
(WI.SpreadsheetCSSStyleDeclarationSection.prototype._handleSpreadsheetSelectorFieldWillNavigate):
(WI.SpreadsheetCSSStyleDeclarationSection.prototype._handleSpreadsheetGroupingFieldDidCommit):
(WI.SpreadsheetCSSStyleDeclarationSection.prototype._handleSpreadsheetGroupingFieldWillNavigate):
(WI.SpreadsheetCSSStyleDeclarationSection.prototype._handleSpreadsheetGroupingFieldDidDiscard):

* Source/WebInspectorUI/UserInterface/Views/SpreadsheetRuleHeaderField.js: Renamed from Source/WebInspectorUI/UserInterface/Views/SpreadsheetSelectorField.js.
(WI.SpreadsheetRuleHeaderField):
(WI.SpreadsheetRuleHeaderField.prototype.get element):
(WI.SpreadsheetRuleHeaderField.prototype.startEditing):
(WI.SpreadsheetRuleHeaderField.prototype.stopEditing):
(WI.SpreadsheetRuleHeaderField.prototype._handleBlur):
(WI.SpreadsheetRuleHeaderField.prototype._handleKeyDown):
- Delegate methods should provide the current text field as their first parameter to allow the delegate to determine the
appropriate action.

* Source/WebInspectorUI/UserInterface/Views/StyleOriginView.js:
(WI.StyleOriginView.prototype.update):
(WI.StyleOriginView):

Canonical link: <a href="https://commits.webkit.org/256043@main">https://commits.webkit.org/256043@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/666b41965bc363291b79c24d0cc64de192c875e3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/94362 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/3540 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/27288 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/104030 "Built successfully") | [❌ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/164313 "Hash 666b4196 for PR 5566 does not build (failure)") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/98360 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/3587 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/31746 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/86699 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/100029 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/100032 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/2573 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/80757 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/29620 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/84495 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/84103 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/72511 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/85579 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/38141 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/17929 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/80734 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/36018 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/19201 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/27757 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4176 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/39905 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/41816 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/83400 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/41855 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/38430 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/18833 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/JavaScriptCore/inspector/protocol/CSS.json:
```diff
@@ -243,8 +243,10 @@
                 { "name": "text", "type": "string", "optional": true, "description": "Query text if specified by a @media, @supports, or @container rule. Layer name (or not present for anonymous layers) for @layer rules." },
-                { "name": "sourceURL", "type": "string", "optional": true, "description": "URL of the document containing the CSS grouping." }
+                { "name": "sourceURL", "type": "string", "optional": true, "description": "URL of the document containing the CSS grouping." },
+                { "name": "sourceRange", "$ref": "SourceRange", "optional": true, "description": "@-rule's header text range in the enclosing stylesheet (if available). This is from the first non-whitespace character after the @ declarartion to the last non-whitespace character before an opening curly bracket or semicolon." }
```
NIT: most other things that use `SourceRange` just call it `range`.  should we match the naming for consistency?

dcrousso commented on Source/JavaScriptCore/inspector/protocol/CSS.json:
```diff
@@ -402,6 +404,18 @@
+            "description": "Modifies an @rule grouping's header text.",
+            "targetTypes": ["page"],
+            "parameters": [
+                { "name": "ruleId", "$ref": "CSSRuleId" },
```
how does this handle scenarios where the same `@` is shared by multiple rules?
```
@media screen {
    @media (min-width: 100px) {
        body {
            ...
        }
    }

    @media (min-width: 200px) {
        body {
            ...
        }
    }
}
```

id hope that modifying the `@media screen` would affect both rulesets

perhaps we need to introduce a `GroupingId`?

regardless, please make sure to add a test case for this :)

patrickangle replied:
This works in the commits from today - The RuleId is a specific enough identifier in this case, and we now share CSSGrouping model objects within a DOMNodeStyle object since they can be displayed together. This means that the single representation of the group for the current DOMNodeStyle will be updated, and will in turn notify its clients of the update, resulting in the text being updated for other occurrences of the same rule. When switching to another node to view its styles, this update is automatic because a style sheet change will have occurred, marking the DOMNodeStyle as dirty, and it will be refreshed before display.

dcrousso replied:
nice!  do we have tests for this exact scenario? 😃

patrickangle replied:
I'll add some!

dcrousso commented on Source/WebCore/css/parser/CSSParser.cpp:
```diff
@@ -94,6 +96,45 @@ bool CSSParser::parseSupportsCondition(const String& condition)
+    CSSParserImpl parser(m_context, condition);
+    if (!parser.tokenizer())
+        return false;
+    return CSSSupportsParser::supportsCondition(parser.tokenizer()->tokenRange(), parser, CSSSupportsParser::ForAtRule) != CSSSupportsParser::Invalid;
```
NIT: you pull out the `tokenRange` in the other methods below.  should we do that here too for consistency?

dcrousso commented on Source/WebCore/css/parser/CSSParserImpl.h:
```diff
@@ -97,6 +98,9 @@ class CSSParserImpl {
     static void parseStyleSheet(const String&, const CSSParserContext&, StyleSheetContents&);
     static CSSSelectorList parsePageSelector(CSSParserTokenRange, StyleSheetContents*);
 
+    enum class AllowAnonymous { Yes, No };
```
NIT: normally we have `No` as the first value 🤔  i realize that this is existing code, but i wonder if we could/should fix this (or at least comment why it has to be this way) while you're at it (or as a followup)

dcrousso commented on Source/WebCore/css/parser/CSSParserImpl.cpp:
```diff
@@ -948,6 +947,8 @@ RefPtr<StyleRuleContainer> CSSParserImpl::consumeContainerRule(CSSParserTokenRan
     if (prelude.atEnd())
         return nullptr;
 
+    auto preludeCopy = prelude;
```
NIT: this could use a better name to clarify why it needs to exist at all (e.g. `originalPreludeLocation`).  alternatively maybe you could make a `CSSParserTokenRange::isWhitespace` that's not destructive and use that instead

dcrousso commented on Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
```diff
@@ -715,17 +708,43 @@ Protocol::ErrorStringOr<Ref<Protocol::CSS::CSSRule>> InspectorCSSAgent::setRuleS
+{
+    Protocol::ErrorString errorString;
+
+    InspectorCSSId compoundId(ruleId);
```
NIT: `WTFMove(ruleId)`

dcrousso commented on Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
```diff
@@ -715,17 +708,43 @@ Protocol::ErrorStringOr<Ref<Protocol::CSS::CSSRule>> InspectorCSSAgent::setRuleS
+    InspectorCSSId compoundId(ruleId);
+    ASSERT(!compoundId.isEmpty());
+
+    InspectorStyleSheet* inspectorStyleSheet = assertStyleSheetForId(errorString, compoundId.styleSheetId());
```
NIT: `auto*`

dcrousso commented on Source/WebCore/inspector/InspectorStyleSheet.h:
```diff
@@ -157,7 +157,8 @@ class InspectorStyleSheet : public RefCounted<InspectorStyleSheet> {
     };
 
-    typedef HashMap<CSSStyleDeclaration*, RefPtr<InspectorStyle>> InspectorStyleMap;
+    typedef std::variant<CSSStyleDeclaration*, CSSRule*> StyleDeclarationOrCSSRule;
```
```
using StyleDeclarationOrCSSRule = std::variant<CSSStyleDeclaration*, CSSRule*>;
```

dcrousso commented on Source/WebCore/inspector/InspectorStyleSheet.h:
```diff
@@ -168,32 +169,34 @@ class InspectorStyleSheet : public RefCounted<InspectorStyleSheet> {
     RefPtr<Inspector::Protocol::CSS::CSSStyleSheetBody> buildObjectForStyleSheet();
     RefPtr<Inspector::Protocol::CSS::CSSStyleSheetHeader> buildObjectForStyleSheetInfo();
     RefPtr<Inspector::Protocol::CSS::CSSRule> buildObjectForRule(CSSStyleRule*);
     Ref<Inspector::Protocol::CSS::CSSStyle> buildObjectForStyle(CSSStyleDeclaration*);
```
Aside: odd that this is the only one that returns a `Ref` 🤔

dcrousso commented on Source/WebCore/inspector/InspectorStyleSheet.h:
```diff
@@ -202,10 +205,10 @@ class InspectorStyleSheet : public RefCounted<InspectorStyleSheet> {
 
 private:
-    typedef Vector<RefPtr<CSSStyleRule>> CSSStyleRuleVector;
+    typedef Vector<RefPtr<CSSRule>> CSSRuleVector;
```
```
using CSSRuleVector = Vector<RefPtr<CSSRule>>;
```

NIT: is this really even necessary to have exist?

dcrousso commented on Source/WebCore/inspector/InspectorStyleSheet.cpp:
```diff
@@ -70,8 +70,92 @@
+enum class RuleFlatteningStrategy {
+    Commit,
+    CommitSelfThenChildren,
+    Ignore,
```
NIT: we normally put the "negative"/default option as the first value in `enum`

dcrousso commented on Source/WebCore/inspector/InspectorStyleSheet.cpp:
```diff
@@ -70,8 +70,92 @@
+        return parser.canParseCascadeLayerName(headerText);
+    case StyleRuleType::Container:
+        return parser.canParseContainerQuery(headerText);
+
```
Style: i don't think we put an extra newline before `default:` 🤔

dcrousso commented on Source/WebCore/inspector/InspectorStyleSheet.cpp:
```diff
@@ -70,8 +70,92 @@
+        return Protocol::CSS::Grouping::Type::LayerRule;
+    case StyleRuleType::Container:
+        return Protocol::CSS::Grouping::Type::ContainerRule;
+
```
ditto (:137)

dcrousso commented on Source/WebCore/inspector/InspectorStyleSheet.cpp:
```diff
@@ -466,82 +519,88 @@ static RefPtr<CSSRuleList> asCSSRuleList(CSSRule* rule)
-static Ref<JSON::ArrayOf<Protocol::CSS::Grouping>> buildArrayForGroupings(CSSRule& rule)
+static String sourceURLForCSSRule(CSSRule& rule)
+{
+    String sourceURL;
```
NIT: I think you can clean this up a bit
```
if (auto* parentStyleSheet = rule.parentStyleSheet()) {
    if (auto sourceURL = parentStyleSheet->contents().baseURL().string(); !sourceURL.isEmpty())
        return sourceURL;

    if (auto* ownerDocument = parentStyleSheet->ownerDocument())
        return InspectorDOMAgent::documentURLString(ownerDocument);
}

return nullString();
```

dcrousso commented on Source/WebCore/inspector/InspectorStyleSheet.cpp:
```diff
@@ -466,82 +519,88 @@ static RefPtr<CSSRuleList> asCSSRuleList(CSSRule* rule)
+    if (sourceData)
+        text = m_parsedStyleSheet->text().substring(sourceData->ruleHeaderRange.start, sourceData->ruleHeaderRange.length());
+
+    if (!text.isEmpty())
```
doesn't this mean that the entire thing could be inside the `if`?
```
if (RefPtr<CSSRuleSourceData> sourceData = ensureParsedDataReady() ? ruleSourceDataFor(groupingRule) : nullptr) {
    if (auto text = text = m_parsedStyleSheet->text().substring(sourceData->ruleHeaderRange.start, sourceData->ruleHeaderRange.length()); !text.isEmpty())
        payload->setText(text);
        
    if (auto range = buildSourceRangeObject(sourceData->ruleHeaderRange, lineEndings()))
        payload->setSourceRange(range.releaseNonNull());
}
```

dcrousso commented on Source/WebCore/inspector/InspectorStyleSheet.cpp:
```diff
@@ -466,82 +519,88 @@ static RefPtr<CSSRuleList> asCSSRuleList(CSSRule* rule)
+        if (auto groupingRule = buildObjectForGrouping(parentRule))
+            groupingsPayload->addItem(groupingRule.releaseNonNull());
+        else if (auto* importRule = dynamicDowncast<CSSImportRule>(parentRule)) {
+            // FIXME: <webkit.org/b/abcxyz> Show import rule as a single rule, instead of two rules for media and layer.
```
oops

dcrousso commented on Source/WebCore/inspector/InspectorStyleSheet.cpp:
```diff
@@ -466,82 +519,88 @@ static RefPtr<CSSRuleList> asCSSRuleList(CSSRule* rule)
+            // FIXME: <webkit.org/b/abcxyz> Show import rule as a single rule, instead of two rules for media and layer.
+            auto sourceURL = sourceURLForCSSRule(*importRule);
+            auto layerName = importRule->layerName();
             if (!layerName.isNull()) {
```
```
if (auto layerName = importRule->layerName(); !layerName.isNull()) {
```

dcrousso commented on Source/WebCore/inspector/InspectorStyleSheet.cpp:
```diff
@@ -466,82 +519,88 @@ static RefPtr<CSSRuleList> asCSSRuleList(CSSRule* rule)
                 mediaRulePayload->setText(media.mediaText());
-                ruleGroupingPayloads.append(WTFMove(mediaRulePayload));
-            }
-        } else if (is<CSSSupportsRule>(parentRule)) {
```
do all of these get properly handled by `buildObjectForGrouping`?  i.e. the `ruleHeaderRange` stuff is the same result?

patrickangle replied:
There are slight variations in the result because the implementations I'm removing below use a serialization of the grouping rule instead of the raw text, which means things like `(min-width:300px)` were previously expanded to `(min-width: 300px)` (note the extra space). However now that we allow editing it feels prudent to provide rule text that matches what the author originally typed instead of the canonical serialization. But yes, these four rule types (Media, Supports, LayerBlock, and Container) are handled by `buildObjectForGrouping` now.

dcrousso commented on Source/WebCore/inspector/InspectorStyleSheet.cpp:
```diff
@@ -552,6 +611,7 @@ static Ref<JSON::ArrayOf<Protocol::CSS::Grouping>> buildArrayForGroupings(CSSRul
         auto* styleSheet = parentRule->parentStyleSheet();
         while (styleSheet) {
             auto* media = styleSheet->media();
+            // FIXME: <webkit.org/b/abcxyz> Support editing `style` and `link` node media queries.
```
oops

dcrousso commented on Source/WebCore/inspector/InspectorStyleSheet.cpp:
```diff
@@ -1377,23 +1459,41 @@ RefPtr<CSSRuleSourceData> InspectorStyleSheet::ruleSourceDataFor(CSSStyleDeclara
 
+RefPtr<CSSRuleSourceData> InspectorStyleSheet::ruleSourceDataFor(CSSRule* rule) const
+{
+    constexpr auto combineSplitRules = true;
```
i think this should really be `static constexpr auto` 🤔

dcrousso commented on Source/WebCore/inspector/InspectorStyleSheet.cpp:
```diff
@@ -1377,23 +1459,41 @@ RefPtr<CSSRuleSourceData> InspectorStyleSheet::ruleSourceDataFor(CSSStyleDeclara
-        if (&rule->style() == pageStyle)
+        auto* cssStyleRule = dynamicDowncast<CSSStyleRule>(rule.get());
+
+        if (WTF::switchOn(ruleOrDeclaration,
```
i'd move this out into a separate variable for easier reading
```
auto matches = WTF::switchOn(
    ...
);
if (matches)
    return index;
```

dcrousso commented on Source/WebCore/inspector/InspectorStyleSheet.cpp:
```diff
@@ -1577,14 +1674,23 @@ void InspectorStyleSheet::collectFlatRules(RefPtr<CSSRuleList>&& ruleList, CSSSt
+
+            auto childRuleList = asCSSRuleList(rule);
+            ASSERT(childRuleList);
+            collectFlatRules(WTFMove(childRuleList), result);
```
should this really be a `WTFMove`?

patrickangle replied:
It previously was, but looking a bit deeper that seems suspicious since we are being handed a reference to the `m_ruleListCSSOMWrapper` of `CSSGroupingRule`...

dcrousso commented on Source/WebInspectorUI/UserInterface/Models/CSSGrouping.js:
```diff
@@ -25,23 +25,48 @@
+    constructor(nodeStyles, ownerStyleSheet, {id, type, text, sourceCodeLocation} = {})
     {
+        console.assert(nodeStyles);
         console.assert(Object.values(CSSGrouping.Type).includes(type));
```
it seems like the `type` is required, so that should not be in the `options = {}`

patrickangle replied:
Oops, still need to do this.

dcrousso commented on Source/WebInspectorUI/UserInterface/Models/CSSGrouping.js:
```diff
@@ -25,23 +25,48 @@
         console.assert(!sourceCodeLocation || sourceCodeLocation instanceof WI.SourceCodeLocation);
 
+        this._nodeStyles = nodeStyles;
+        this._ownerStyleSheet = ownerStyleSheet || null;
```
if this is not actually required then it should be in the `options = {}`

patrickangle replied:
ditto

dcrousso commented on Source/WebInspectorUI/UserInterface/Models/CSSGrouping.js:
```diff
@@ -25,23 +25,48 @@
+        return this._text;
+    }
+
+    setText(newText)
```
did you mean `set text(text)`?

patrickangle replied:
I did not, since this is effectively `async` (and is actually `async` on my the upcoming commit), and we need to always afterwards update the text in case the backend rejected the change, but we can't preemptively update the text because we would flash to the old value while we wait for the backend to eventually get back to us so we can update the text again.

dcrousso commented on Source/WebInspectorUI/UserInterface/Models/CSSGrouping.js:
```diff
@@ -80,6 +105,31 @@ WI.CSSGrouping = class CSSGrouping
+    {
+        console.assert(groupingPayload.type == this._type);
+
+        console.assert(groupingPayload.ruleId);
```
since this is coming from the protocol, it shouldn't be necessary to assert that these things exist

patrickangle replied:
These are technically optional in the protocol, and could indicate something has gone wrong if we lack a RuleId or text after the change.

dcrousso commented on Source/WebInspectorUI/UserInterface/Models/DOMNodeStyles.js:
```diff
@@ -459,6 +459,37 @@ WI.DOMNodeStyles = class DOMNodeStyles extends WI.Object
+        function ruleSelectorChanged(error, rulePayload)
+        {
+            if (error) {
+                console.log(error);
```
oops

dcrousso commented on Source/WebInspectorUI/UserInterface/Models/DOMNodeStyles.js:
```diff
@@ -459,6 +459,37 @@ WI.DOMNodeStyles = class DOMNodeStyles extends WI.Object
 
+    changeGroupingText(grouping, text)
+    {
+        text = text || "";
```
```
text ||= ""
```

dcrousso commented on Source/WebInspectorUI/UserInterface/Models/DOMNodeStyles.js:
```diff
@@ -459,6 +459,37 @@ WI.DOMNodeStyles = class DOMNodeStyles extends WI.Object
         return result.promise;
     }
 
+    changeGroupingText(grouping, text)
```
i think you can make this cleaner using `async`
```
async changeGroupingText(grouping, text)
{
    text ||= "";
    
    this._needsRefresh = true;
    this.ignoreNextContentDidChangeForStyleSheet = grouping.ownerStyleSheet;

    let target = WI.assumingMainTarget();
    let {grouping: groupingPayload} = await target.setGroupingHeaderText(grouping.id, text);
    
    target.DOMAgent.markUndoableState();
    await this.refresh();

    return groupingPayload;
}
```

also, why not have this be on `WI.CSSGrouping` instead of `WI.DOMNodeStyles`?  you're passing the `WI.CSSGrouping` into this method anyways, so it seems like that's really the focus here, not the `WI.DOMNodeStyles` (which is really only used for the `refresh`)

dcrousso commented on Source/WebInspectorUI/UserInterface/Models/CSSGrouping.js:
```diff
@@ -80,6 +105,31 @@ WI.CSSGrouping = class CSSGrouping
+        }
+
+        let groupingSourceCodeLocation = WI.DOMNodeStyles.createSourceCodeLocation(groupingPayload.sourceURL, location);
+        groupingSourceCodeLocation = WI.cssManager.styleSheetForIdentifier(this._id.styleSheetId).offsetSourceCodeLocation(groupingSourceCodeLocation);
```
NIT: id prefer if we didn't reassign the same variable, as that can make it harder to debug things.  i'd inline this in the `this._sourceCodeLocation = ...` instead

dcrousso commented on Source/WebInspectorUI/UserInterface/Models/DOMNodeStyles.js:
```diff
@@ -712,10 +743,24 @@ WI.DOMNodeStyles = class DOMNodeStyles extends WI.Object
+
+            let groupingSourceCodeLocation = WI.DOMNodeStyles.createSourceCodeLocation(grouping.sourceURL, location);
             if (styleSheet)
                 groupingSourceCodeLocation = styleSheet.offsetSourceCodeLocation(groupingSourceCodeLocation);
```
ditto ([Source/WebInspectorUI/UserInterface/Models/CSSGrouping.js](#diff-09c45dc733ff4be6e27d1fe184904524721db668ad98087bbb46e61b2bb5f267):129)

patrickangle replied:
I agree on CSSGrouping.js:129, but disagree here. The alternative here seems much harder to grok in my opinion:
```
let groupingSourceCodeLocation = WI.DOMNodeStyles.createSourceCodeLocation(grouping.sourceURL, location);
let offsetGroupingSourceCodeLocation;
if (styleSheet)
    offsetGroupingSourceCodeLocation = styleSheet.offsetSourceCodeLocation(groupingSourceCodeLocation);
```
And then having to use `offsetGroupingSourceCodeLocation ?? groupingSourceCodeLocation`.

dcrousso replied:
id use a ternary for this
```
let groupingSourceCodeLocation = WI.DOMNodeStyles.createSourceCodeLocation(grouping.sourceURL, location);
let offsetGroupingSourceCodeLocation = styleSheet ? styleSheet.offsetSourceCodeLocation(groupingSourceCodeLocation) : groupingSourceCodeLocation;
```

or even just an optional chaining with nullish coalescing combo :)

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SpreadsheetCSSStyleDeclarationSection.css:
```diff
@@ -141,11 +141,11 @@
 }
 
-.spreadsheet-css-declaration .selector.spreadsheet-selector-field {
+.spreadsheet-css-declaration :matches(.selector, .grouping).spreadsheet-rule-header-field {
```
`:is` instead of `:matches`

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SpreadsheetCSSStyleDeclarationSection.css:
```diff
@@ -141,11 +141,11 @@
 }
 
-.spreadsheet-css-declaration .selector.spreadsheet-selector-field.editing {
+.spreadsheet-css-declaration :matches(.selector, .grouping).spreadsheet-rule-header-field.editing {
```
ditto (:144)

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SpreadsheetCSSStyleDeclarationSection.js:
```diff
@@ -107,22 +107,29 @@ WI.SpreadsheetCSSStyleDeclarationSection = class SpreadsheetCSSStyleDeclarationS
+
+                if (grouping.editable) {
+                    let groupingTextField = new WI.SpreadsheetRuleHeaderField(this, groupingTextElement);
+                    groupingTextField.addEventListener(WI.SpreadsheetRuleHeaderField.Event.StartedEditing, function(event) {
```
these should really be member functions so that we don't have to create a new instance of the same `function` each time we render a selector

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SpreadsheetCSSStyleDeclarationSection.js:
```diff
@@ -107,22 +107,29 @@ WI.SpreadsheetCSSStyleDeclarationSection = class SpreadsheetCSSStyleDeclarationS
+                    groupingTextField.addEventListener(WI.SpreadsheetRuleHeaderField.Event.StartedEditing, function(event) {
+                        this._headerElement.classList.add("editing-selector");
+                    }, this);
+                    groupingTextField.addEventListener(WI.SpreadsheetRuleHeaderField.Event.StoppedEditing, function(event) {
```
ditto (:121)

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SpreadsheetCSSStyleDeclarationSection.js:
```diff
@@ -144,11 +151,11 @@ WI.SpreadsheetCSSStyleDeclarationSection = class SpreadsheetCSSStyleDeclarationS
-            this._selectorTextField = new WI.SpreadsheetSelectorField(this, this._selectorElement);
-            this._selectorTextField.addEventListener(WI.SpreadsheetSelectorField.Event.StartedEditing, function(event) {
+            this._selectorTextField = new WI.SpreadsheetRuleHeaderField(this, this._selectorElement);
+            this._selectorTextField.addEventListener(WI.SpreadsheetRuleHeaderField.Event.StartedEditing, function(event) {
```
ditto (:121)

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SpreadsheetCSSStyleDeclarationSection.js:
```diff
@@ -144,11 +151,11 @@ WI.SpreadsheetCSSStyleDeclarationSection = class SpreadsheetCSSStyleDeclarationS
                 this._headerElement.classList.add("editing-selector");
             }, this);
-            this._selectorTextField.addEventListener(WI.SpreadsheetSelectorField.Event.StoppedEditing, function(event) {
+            this._selectorTextField.addEventListener(WI.SpreadsheetRuleHeaderField.Event.StoppedEditing, function(event) {
```
ditto (:121)

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SpreadsheetCSSStyleDeclarationSection.js:
```diff
@@ -694,6 +719,84 @@ WI.SpreadsheetCSSStyleDeclarationSection = class SpreadsheetCSSStyleDeclarationS
+        }
+
+        this.dispatchEventToListeners(WI.SpreadsheetCSSStyleDeclarationSection.Event.SelectorWillChange);
+        groupingTextElement.representedGrouping.setText(text).catch((error) => { }).finally(() => {
```
i don think the `.catch((error) => { })` is necessary since you have the `finally` 🤔

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SpreadsheetCSSStyleDeclarationSection.js:
```diff
@@ -694,6 +719,84 @@ WI.SpreadsheetCSSStyleDeclarationSection = class SpreadsheetCSSStyleDeclarationS
+            return;
+        }
+
+        this.dispatchEventToListeners(WI.SpreadsheetCSSStyleDeclarationSection.Event.SelectorWillChange);
```
should we create a `GroupingWillChange` event so we're not overloading this one?

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SpreadsheetRuleHeaderField.js:
```diff
@@ -110,9 +111,9 @@ WI.SpreadsheetSelectorField = class SpreadsheetSelectorField extends WI.Object
             return;
 
-        if (this._delegate && typeof this._delegate.spreadsheetSelectorFieldDidCommit === "function") {
+        if (typeof this._delegate?.spreadsheetRuleHeaderFieldDidCommit === "function") {
```
NIT: you can also drop the `typeof` part too.  we don't support the possibility that a property with the same name as a delegate method be anything other than a function

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SpreadsheetRuleHeaderField.js:
```diff
@@ -133,9 +134,9 @@ WI.SpreadsheetSelectorField = class SpreadsheetSelectorField extends WI.Object
             event.stop();
 
-            if (this._delegate && typeof this._delegate.spreadsheetSelectorFieldWillNavigate === "function") {
+            if (typeof this._delegate?.spreadsheetRuleHeaderFieldWillNavigate === "function") {
```
ditto (:114)

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SpreadsheetRuleHeaderField.js:
```diff
@@ -146,13 +147,13 @@ WI.SpreadsheetSelectorField = class SpreadsheetSelectorField extends WI.Object
 
-            if (this._delegate && typeof this._delegate.spreadsheetSelectorFieldDidDiscard === "function")
-                this._delegate.spreadsheetSelectorFieldDidDiscard();
+            if (typeof this._delegate?.spreadsheetRuleHeaderFieldDidDiscard === "function")
```
ditto (:114)

dcrousso reviewed (commented):
overall this looks *really* good! 😃

mostly Style/NIT, but i do have a few legit questions/concerns that i'd like to see before i approve

dcrousso commented on Source/JavaScriptCore/inspector/protocol/CSS.json:
```diff
@@ -243,8 +243,10 @@
             "description": "CSS @media (as well as other users of media queries, like @import, <style>, <link>, etc.), @supports, and @layer descriptor.",
             "properties": [
                 { "name": "type", "type": "string", "enum": ["media-rule", "media-import-rule", "media-link-node", "media-style-node", "supports-rule", "layer-rule", "layer-import-rule", "container-rule"], "description": "Source of the media query: \"media-rule\" if specified by a @media rule, \"media-import-rule\" if specified by an @import rule, \"media-link-node\" if specified by a \"media\" a ...
+                { "name": "ruleId", "$ref": "CSSRuleId", "optional": true, "description": "The CSS rule identifier (absent for non-editable grouping rules)."},
```
should we create a `CSSGroupingId` just so that it's a unique type?

or is the explicit idea here that we use the same `CSSRuleId` as the related `CSSRule`?  if so, can we indicate that somewhere in the `description`?

patrickangle replied:
I'm hesitant to create a new third form of the ID for what is at the end of the day the same thing. We already have CSSStyleId and CSSRuleId, both of which represent the same thing, a style sheet and rule ordinal, just one of them is only ever used for Style rules. 

Conceptually all of the @rules are `CSSGroupingRule`, which is a type of CSSRule.

I'll update the description of the property to make this clear.

dcrousso commented on Source/WebInspectorUI/UserInterface/Models/CSSGrouping.js:
```diff
@@ -23,25 +23,80 @@
+        let groupingSourceCodeLocation = WI.DOMNodeStyles.createSourceCodeLocation(groupingPayload.sourceURL, location);
+        this._sourceCodeLocation = WI.cssManager.styleSheetForIdentifier(this._id.styleSheetId).offsetSourceCodeLocation(groupingSourceCodeLocation);
+
+        this.dispatchEventToListeners(WI.CSSGrouping.Event.GroupingUpdated);
```
NIT: i kinda feel like we should have this be named more specific, like `WI.CSSGrouping.Event.TextChanged` and/or` WI.CSSGrouping.Event.LocationChanged` instead of this very generic thing 🤔

dcrousso commented on Source/WebInspectorUI/UserInterface/Models/DOMNodeStyles.js:
```diff
@@ -711,11 +715,48 @@ WI.DOMNodeStyles = class DOMNodeStyles extends WI.Object
 
         // COMPATIBILITY (iOS 13): CSS.CSSRule.groupings did not exist yet.
         let groupings = (payload.groupings || payload.media || []).map((grouping) => {
+            let ruleId = grouping.ruleId;
```
i think we also should have a `// COMPATIBILITY` for this

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SpreadsheetCSSStyleDeclarationSection.js:
```diff
@@ -318,6 +310,47 @@ WI.SpreadsheetCSSStyleDeclarationSection = class SpreadsheetCSSStyleDeclarationS
+        } else
+            this._groupingsContainerElement.removeChildren();
+
+        let groupings = this._style.groupings.reverse();
```
this is an in-place modification, so it will flip the order each time it's called

i think we either need to `.slice().reverse()` or just do a `for` that goes from `.length - 1` to `0`

patrickangle replied:
Oh good catch! We used to filter the results before the reverse, but now we don't. Given we don't need to filter I think just iterating by index will work just fine.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SpreadsheetCSSStyleDeclarationSection.js:
```diff
@@ -694,6 +727,103 @@ WI.SpreadsheetCSSStyleDeclarationSection = class SpreadsheetCSSStyleDeclarationS
+
+        this.dispatchEventToListeners(WI.SpreadsheetCSSStyleDeclarationSection.Event.SelectorWillChange);
+        groupingTextElement.representedGrouping.setText(text).catch((error) => {
+            console.error(error);
```
is this necessary?  it should already be caught by the uncaught exception reporter.  also if truly so, please use `WI.reportInternalError`

patrickangle replied:
This isn't necessary, no. I was debugging some bad syntax of mine earlier 😅

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SpreadsheetCSSStyleDeclarationSection.js:
```diff
@@ -318,6 +310,47 @@ WI.SpreadsheetCSSStyleDeclarationSection = class SpreadsheetCSSStyleDeclarationS
         this._renderSelector();
     }
 
+    _layoutGroupings()
```
NIT: i think `_createGroupingElements` would be more apt :)

dcrousso commented on Source/WebCore/inspector/InspectorStyleSheet.cpp:
```diff
@@ -70,8 +72,122 @@
+        // Make sure the engine can parse the provided `@` rule, even if it only uses unsupported features. As long as
+        // the rule text is entirely consumed and it creates a rule of the expected type, we consider it valid because
+        // we will be able to continue to edit the rule in the future.
+        CSSParserImpl parser(parserContextForDocument(document), atRuleIdentifier + ' ' + headerText + " {}");
```
clever!

dcrousso commented on LayoutTests/inspector/css/generateFormattedText-expected.txt:
```diff
@@ -50,8 +50,8 @@ a: 1; b: 2; c: 3;
 a: 1; b: 2; c: 3;
 
 -- Running test case: CSS.generateFormattedText.MatchedRules.includeGroupingsAndSelectors
-@media only screen and (min-width: 0px) {@media only screen and (min-height: 0px) {body > div#test-node {a: 1; b: 2; c: 3;}}}
```
odd that this changed 🤔

i wouldn't expect `WI.CSSFormatter` to be affected by any changes to `WI.CSSGrouping` (since IIRC it runs in a `Worker`).  any idea why this happened, and maybe a way to prevent that?

patrickangle replied:
These tests test `WI.CSSStyleDeclaration.generateFormattedText`, which uses the text from the CSSGrouping. The only time this text is used is for copying a rule to the clipboard. These slight changes exist because we no longer present the CSSOM's serialized value for grouping, but rather present the authored text, which in this case does not contain these spaces between the `:` and the value.

dcrousso replied:
ooohhh oops this is `WI.CSSStyleDeclaration.prototype.generateFormattedText`, not `WI.CSSFormatter`

yeah this is totally fine to change then :)

dcrousso reviewed (approved):
nice work!  a few remaining things to add/fix :)

patrickangle committed:
Web Inspector: Support editing @rules in the Styles sidebar
https://bugs.webkit.org/show_bug.cgi?id=246768
rdar://99871652

Reviewed by Devin Rousso.

This work is best thought of in three interlocking pieces:
- Rework chunks of InspectorStyleSheet to allow us to determine source data for @rules, including the header range,
which is the range of the text between `@media` (and friends) and the opening curly bracket for the rule set.
- Generalizing setting a rule's selector to just be setting its header text so we can do so with non-style rules.
- Implement frontend support for editing these rules in the same way selectors are currently editable.

* LayoutTests/inspector/css/setGroupingHeaderText-expected.txt: Added.
* LayoutTests/inspector/css/setGroupingHeaderText.html: Added.

* LayoutTests/inspector/css/generateFormattedText-expected.txt:
* LayoutTests/inspector/css/getMatchedStylesForNode-expected.txt:
- Rebaseline expectations for new CSS.Grouping properties.

* Source/JavaScriptCore/inspector/protocol/CSS.json:
- Add rule ID to CSS.Grouping so that we can use it later to edit the header text of the grouping.
- Add method CSS.setGroupingHeaderText to change a CSS.Grouping's header text.

* Source/WebCore/css/parser/CSSParser.h:
* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::CSSParserImpl::consumeContainerRule):
- Provide a correct header range by keeping a copy of the original prelude before consumption to use for informing the
observer.

* Source/WebCore/css/parser/CSSParserImpl.h:
- Expose `consumeAtRule` publicly for use in InspectorStyleSheet

* Source/WebCore/inspector/InspectorStyleSheet.cpp:
(WebCore::flatteningStrategyForStyleRuleType):
- Move the logic for determining how we unflatten rules to a shared static method so that the source data and flat CSSOM
rule lists always consider the same set of rules in the same way.

(WebCore::parserContextForDocument):
(WebCore::isValidRuleHeaderText):
(WebCore::protocolGroupingTypeForStyleRuleType):
- Abstract logic that relies on rule type into static methods so that the various bits that need implemented when adding
an editable rule type are in close proximity to each other.

(WebCore::flattenSourceData):
(WebCore::ParsedStyleSheet::setSourceData):
- Use `flatteningStrategyForStyleRuleType` to determine how we should handle each piece of source data.

(WebCore::StyleSheetHandler::startRuleHeader):
(WebCore::StyleSheetHandler::setRuleHeaderEnd):
- Fix to never produce a range end that occurs before the range's start.

(WebCore::sourceURLForCSSRule):
(WebCore::InspectorStyleSheet::buildObjectForGrouping):
(WebCore::InspectorStyleSheet::buildArrayForGroupings):
- Split building the grouping array into a few discrete, reusable pieces. For the four types of rules we are making
editable, we will use `buildObjectForGrouping` to build those rule objects. For other real that are currently turned
into fake groupings, we leave the logic mostly intact (which means they are not editable).

(WebCore::InspectorStyleSheet::ruleHeaderText):
(WebCore::InspectorStyleSheet::setRuleHeaderText):
(WebCore::InspectorStyleSheet::ruleSelector): Deleted.
(WebCore::isValidSelectorListString): Deleted.
(WebCore::InspectorStyleSheet::setRuleSelector): Deleted.
- Setting a rule's header text and setting a rule's selector(s) are (except for some optimizations) the same operation.
For CSSStyleRule we still follow the existing logic of modifying a selector via CSSOM, which saves us from reparsing the
entire style sheet. For other rules, this is more complicated because our CSSOM implementation doesn't allow mutating
the header text, and changing these rules could change which rules apply to which nodes beyond just the rule we are
mutating.

(WebCore::InspectorStyleSheet::deleteRule):
(WebCore::InspectorStyleSheet::ruleForId const):
(WebCore::InspectorStyleSheet::cssStyleRulesSplitFromSameRule):
(WebCore::InspectorStyleSheet::buildObjectForSelectorList):
(WebCore::InspectorStyleSheet::buildObjectForRule):
(WebCore::InspectorStyleSheet::styleForId const):
(WebCore::InspectorStyleSheet::ruleOrStyleId const):
(WebCore::InspectorStyleSheet::ruleSourceDataFor const):
(WebCore::InspectorStyleSheet::ruleIndexByStyle const):
(WebCore::InspectorStyleSheet::buildArrayForRuleList):
(WebCore::InspectorStyleSheet::collectFlatRules):
(WebCore::InspectorStyleSheet::ruleId const): Deleted.
- Don't assume that a CSSRule will always be a CSSStyleRule.
- Allow both styles and rules to be used to look up the ID for a rule/style.

* Source/WebCore/inspector/InspectorStyleSheet.h:
(WebCore::InspectorStyleSheet::canBind const):
(WebCore::InspectorStyleSheet::styleId const): Deleted.

* Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
(WebCore::InspectorCSSAgent::setRuleSelector):
(WebCore::InspectorCSSAgent::setGroupingHeaderText):
* Source/WebCore/inspector/agents/InspectorCSSAgent.h:
- Generalize setting selectors to also support setting header text of @rules.

(WebCore::InspectorCSSAgent::addRule):
- Don't assume that all rules are CSSStyleRules.

(WebCore::InspectorCSSAgent::asCSSStyleRule): Deleted.
- Remove helper that was only used in a few places and is better served by a dynamicDowncast anyways.

* Source/WebInspectorUI/UserInterface/Main.html:

* Source/WebInspectorUI/UserInterface/Models/CSSGrouping.js:
(WI.CSSGrouping.prototype.get ownerStyleSheet):
(WI.CSSGrouping.prototype.get id):
(WI.CSSGrouping.prototype.get type):
(WI.CSSGrouping.prototype.get editable):
(WI.CSSGrouping.prototype.get text):
- Add necessary information for editing, which is mostly keeping track of our nodeStyles and ownerStyleSheet like style
rules do, and accepting an optional ID, which when present allows us to perform edits.

(WI.CSSGrouping.prototype.async setText):
- Handle setting the text for the grouping. For groupings that no longer apply after modification, we need to update the
current grouping object so that it is accurate to the current state, as it will remain visible to the user until they
switch to a different node.

* Source/WebInspectorUI/UserInterface/Models/DOMNodeStyles.js:
(WI.DOMNodeStyles):
(WI.DOMNodeStyles.prototype.set ignoreNextContentDidChangeForStyleSheet):
(WI.DOMNodeStyles.prototype.refresh.fetchedMatchedStyles):
(WI.DOMNodeStyles.prototype._parseRulePayload):
- Parse the additional data in the payload to support editing.
- Use a shared representation of CSSGrouping's with an ID so that we can update the common representation during an
edit, thus allowing that same rule to update in all the locations it is visible for the current node. (It is not
necessary to update for other nodes, since they will have been invalidated by the stylesheet change and will reload
their styles.)

* Source/WebInspectorUI/UserInterface/Views/SpreadsheetCSSStyleDeclarationSection.css:
(.spreadsheet-css-declaration :is(.selector, .grouping).spreadsheet-rule-header-field):
(.spreadsheet-css-declaration :is(.selector, .grouping).spreadsheet-rule-header-field.editing):
- Style the grouping editor field the same way we style the selector editor field.

* Source/WebInspectorUI/UserInterface/Views/SpreadsheetCSSStyleDeclarationSection.js:
(WI.SpreadsheetCSSStyleDeclarationSection):
(WI.SpreadsheetCSSStyleDeclarationSection.prototype.initialLayout):
- Remove logic that lays out the CSS groupings, since those are dynamic and may change on future updates.

(WI.SpreadsheetCSSStyleDeclarationSection.prototype.layout):
(WI.SpreadsheetCSSStyleDeclarationSection.prototype._layoutGroupings.this._groupingElements.groupings.map.):
(WI.SpreadsheetCSSStyleDeclarationSection.prototype._layoutGroupings):
- Update groupings when layout is called to get the latest information from DOMNodeStyles.
- No longer merge together adjacent @rules like multiple @media or @layer rules, as those rules are editable now, and we
need to accurately represent the author's rules.
- Associate a text field with the grouping header text when applicable to enable editing.

(WI.SpreadsheetCSSStyleDeclarationSection.prototype.spreadsheetRuleHeaderFieldDidCommit):
(WI.SpreadsheetCSSStyleDeclarationSection.prototype.spreadsheetRuleHeaderFieldWillNavigate):
(WI.SpreadsheetCSSStyleDeclarationSection.prototype.spreadsheetRuleHeaderFieldDidDiscard):
(WI.SpreadsheetCSSStyleDeclarationSection.prototype.spreadsheetSelectorFieldDidCommit): Deleted.
(WI.SpreadsheetCSSStyleDeclarationSection.prototype.spreadsheetSelectorFieldWillNavigate): Deleted.
(WI.SpreadsheetCSSStyleDeclarationSection.prototype.spreadsheetSelectorFieldDidDiscard): Deleted.
- Share the delegate methods for the selector and the grouping fields.

(WI.SpreadsheetCSSStyleDeclarationSection.prototype._handleSpreadsheetSelectorFieldStartedEditing):
(WI.SpreadsheetCSSStyleDeclarationSection.prototype._handleSpreadsheetSelectorFieldStoppedEditing):
(WI.SpreadsheetCSSStyleDeclarationSection.prototype._handleSpreadsheetSelectorFieldDidCommit):
(WI.SpreadsheetCSSStyleDeclarationSection.prototype._handleSpreadsheetSelectorFieldWillNavigate):
(WI.SpreadsheetCSSStyleDeclarationSection.prototype._handleSpreadsheetGroupingFieldDidCommit):
(WI.SpreadsheetCSSStyleDeclarationSection.prototype._handleSpreadsheetGroupingFieldWillNavigate):
(WI.SpreadsheetCSSStyleDeclarationSection.prototype._handleSpreadsheetGroupingFieldDidDiscard):

* Source/WebInspectorUI/UserInterface/Views/SpreadsheetRuleHeaderField.js: Renamed from Source/WebInspectorUI/UserInterface/Views/SpreadsheetSelectorField.js.
(WI.SpreadsheetRuleHeaderField):
(WI.SpreadsheetRuleHeaderField.prototype.get element):
(WI.SpreadsheetRuleHeaderField.prototype.startEditing):
(WI.SpreadsheetRuleHeaderField.prototype.stopEditing):
(WI.SpreadsheetRuleHeaderField.prototype._handleBlur):
(WI.SpreadsheetRuleHeaderField.prototype._handleKeyDown):
- Delegate methods should provide the current text field as their first parameter to allow the delegate to determine the
appropriate action.

* Source/WebInspectorUI/UserInterface/Views/StyleOriginView.js:
(WI.StyleOriginView.prototype.update):
(WI.StyleOriginView):

Canonical link: https://commits.webkit.org/256043@main
